File: C++/Queue/Implementaciones/Queue.cpp
#include "Queue.h"
#define endl "\n"


Node::Node(int val){
    data = val;
    next = nullptr;
}

Queue::Queue(){
    head = nullptr;
    back = nullptr;
}

void Queue::enqueue(int element){
    Node* newNode = new Node(element);
    if (head == nullptr){
        head = newNode;
        back = newNode;
        return;
    }
    else{
        back -> next = newNode;
        back = newNode;
        return;
    }
}

Node* Queue::dequeue(){
    if (head == nullptr)
        return nullptr;
    else{
        Node* temp = head;
        head = head -> next;
        return temp;
    }
}

bool Queue::is_empty(){
    if (head == nullptr)
        return true;
    return false;
}

Node* Queue::front(){
    return head;
}

Node* Queue::rear(){
    return back;
}

void Queue::display(){
    Node* iter = head;
    while(iter != nullptr){
        cout << iter -> data << " ";
        iter = iter -> next;
    }
    cout << endl;
    return;
}

int main(){
    Queue cola;
    cout << "Insertar elementos 1, 2 y 3" << endl;
    cola.enqueue(1);
    cola.enqueue(2);
    cola.enqueue(3);
    cout << "Imprimir la cola" << endl;
    cola.display();
    cout << "Hacer dequeue()" << endl;
    cola.dequeue();
    cout << "Imprimir la cola" << endl;
    cola.display();
    cout << boolalpha;
    cout << "Mirar si la cola esta empty" << endl;
    cout << cola.is_empty() << endl;
    cout << "Mirar el primer elemento de la cola" << endl;
    cout << cola.front() -> data << endl;
    cout << "Mirar el último elemento de la cola" << endl;
    cout << cola.rear() -> data << endl;
}

File: Java/Queue/Implementaciones/Queue.java
package Queue.Implementaciones;

public class Queue<T> {
    private Node<T> head;
    private Node<T> tail;

    public Queue() {
        this.head = null;
        this.tail = null;
    }

    // Métodfo para validar si la cola esta vacia
    public boolean isEmpty() {
        return head == null;
    }

    // Método para agregar un elemento al final de la cola
    public void enqueue(T value) {
        Node<T> newNode = new Node<>(value);
        if (isEmpty()) {
            head = newNode;
            tail = newNode;
        } else {
            tail.next = newNode;
            tail = newNode;
        }
    }

    // Método para eliminar y retornar el elemento al frente de la cola
    public Node<T> dequeue() {
        if (isEmpty()) {
            return null;
        }
        Node<T> temp = head;
        head = head.next;
        if (head == null) {
            tail = null;
        }
        return temp;
    }

    // Metodo para mostrar la lista enlazada
    public static <T> void printList(Node<T> head) {
        Node<T> current = head;
        // recorrer hasta current sea null
        while (current != null) {
            System.out.print(current.data + " ");
            current = current.next;
        }
    }

    // Metodo que devuelve el valor del primer elemento de la cola
    public T front() {
        return head.data;
    }
    // Metodo que devuelve el valor del último elemento de la cola
    public T rear() {
        return tail.data;
    }

    // Metodo para obtener el tamaño de la cola
    public int size() {
        int count = 0;
        Node<T> current = head;
        while (current != null) {
            count++;
            current = current.next;
        }
        return count;
    }

    public static void main(String[] args) {
        Queue<Integer> cola = new Queue<>();
        cola.enqueue(1);
        cola.enqueue(2);
        cola.enqueue(3);

        // imprimir los elementos de la cola
        printList(cola.head);
        System.out.println();
        System.out.println("Primer elemento de la cola: " + cola.front());
        System.out.println("Último elemento de la cola: " + cola.rear());
        // eliminar el primer elemento de la cola
        cola.dequeue();
        printList(cola.head);
    }
}
